onedal/backend/data/data.cpp

```diff
@@ -0,0 +1,168 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <stdexcept>
+#include <string>
+
+#include <Python.h>
+
+#include "backend/data/data.h"
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+#include <CL/sycl.hpp>
+#endif
+
+#ifdef DPCTL_ENABLE
+#include "dpctl_sycl_types.h"
+#include "dpctl_sycl_queue_manager.h"
+#endif
+
+namespace oneapi::dal::python {
+#define is_array(a)         ((a) && PyArray_Check(a))
+#define array_type(a)       PyArray_TYPE((PyArrayObject *)a)
+#define array_is_behaved(a) (PyArray_ISCARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_behaved_F(a) \
+    (PyArray_ISFARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_native(a) (PyArray_ISNOTSWAPPED((PyArrayObject *)a))
+#define array_numdims(a)   PyArray_NDIM((PyArrayObject *)a)
+#define array_data(a)      PyArray_DATA((PyArrayObject *)a)
+#define array_size(a, i)   PyArray_DIM((PyArrayObject *)a, i)
+
+#ifdef ONEDAL_DATA_PARALLEL
+int init_numpy() {
+    import_array();
+    return 0;
+}
+
+const static int numpy_initialized = init_numpy();
+#endif
+
+template <typename T>
+inline dal::homogen_table convert_from_numpy_to_homogen(PyArrayObject *array) {

```

Comment:
It's clear that you are converting from numpy because of the input parameter. Maybe, name it shorter? 
`convert_to_homogen_table`